torchao/prototype/low_bit_optim/cpu_offload.py

```diff
@@ -0,0 +1,101 @@
+        """Offload optimizer to CPU for single-GPU training. This will reduce GPU memory by the size of optimizer state.
+        Optimizer step will be done on CPU.
+
+        Args
+            params: a list of parameters or parameter groups.
+            optimizer_class: constructor of the base optimizer.
+            offload_gradients: free GPU gradients once they are moved to CPU. Not compatible with gradient accumulation.
+            kwargs: other keyword arguments to be passed to the base optimizer e.g. `lr`, `weight_decay`.
+        """
+        param_groups = list(params)
+        if len(param_groups) == 0:
+            raise ValueError("optimizer got an empty parameter list")
+        if not isinstance(param_groups[0], dict):
+            param_groups = [{"params": param_groups}]
+
+        self.param_cuda2cpu_map = dict()
+        self.optim_dict = dict()
+        self.stream = torch.cuda.Stream()
+
+        # the queue maintains the order which param we should do optim step on first.
+        self.queue = dict()
+
+        def backward_hook(p_cuda):
+            if p_cuda.grad is not None:
+                p_cpu = self.param_cuda2cpu_map[p_cuda]
+
+                # make sure backward for this param finishes
+                self.stream.wait_stream(torch.cuda.current_stream())
+                with torch.cuda.stream(self.stream):
+                    p_cpu.grad.copy_(p_cuda.grad, non_blocking=True)
+
+                # we rely on CPython implementation of dictionary, which preserves insertion order.
+                # if a param is added again (e.g. due to gradient accumulation), it is moved to the
+                # end of the queue by removing and inserting it again.
+                if p_cuda in self.queue:
+                    del self.queue[p_cuda]
+                self.queue[p_cuda] = self.stream.record_event()
+
+                # deallocate CUDA gradients once D2H transfer finishes.
+                if offload_gradients:
+                    p_cuda.grad.record_stream(self.stream)
+                    p_cuda.grad = None
+
+        for param_group in param_groups:
+            params = param_group.pop("params")
+
+            for p_cuda in params:
+                p_cpu = p_cuda.detach().cpu().pin_memory()
+                p_cpu.grad = torch.empty_like(p_cpu, pin_memory=True)  # pre-allocate CPU grad
+                self.param_cuda2cpu_map[p_cuda] = p_cpu
+
+                p_cuda.register_post_accumulate_grad_hook(backward_hook)
+                self.optim_dict[p_cuda] = optimizer_class([{"params": p_cpu, **param_group}], **kwargs)
+
+    @torch.no_grad()
+    def step(self, closure=None):
+        loss = None
+        if closure is not None:
+            loss = closure()
+
+        for p_cuda, grad_d2h_event in self.queue.items():
+            grad_d2h_event.synchronize()
+            self.optim_dict[p_cuda].step()
```

Comment:
ah, so fused optimizer only fuses vertically? (or is there a potential perf hit here by running per-parameter fused optimizer step?)

Reply:
there is some perf hit due to calling fused Adam on each parameter separately (my current approach) instead of all (or some) parameters (650ms -> 750ms iirc). I couldn't figure out a way to call fused Adam on more than one parameter because of synchronization: in `__init__()`, we don't know which params will have theirs grads D2H finish first, so we can't statically schedule and group the params.

Technically it's still possible if we use functional Adam (i.e. wait for a few items in `self.queue`, then call functional Adam on them), but then it would require writing optim-specific code, instead of treating base optimizer as a black box.